README.md
mELM [![GitHub version](https://badge.fury.io/gh/MattOates%2Fmelm.svg)](https://badge.fury.io/gh/MattOates%2Fmelm)[![Build Status](https://travis-ci.com/MattOates/melm.svg?branch=master)](https://travis-ci.com/MattOates/melm)[![Coverage Status](https://coveralls.io/repos/github/MattOates/melm/badge.svg?branch=master)](https://coveralls.io/github/MattOates/melm?branch=master)
==================

[⬇️ Download latest release v1.4.3](https://github.com/MattOates/melm/releases/download/v1.4.3/melm)

mELM is a tool for masking or assigning Eukaryotic Linear Motifs to protein sequences. Both TSV/GFF3 output or FASTA is possible. 
Essentially the tool is a CLI to the ELM.eu.org online resource with additional tools for dealing with short motif assignment within disordered regions.
ANCHOR can be used to attempt to limit motif assignment to those found in disordered or MoRF regions.

In v1.4.1 onwards mELM has moved to using a full class library and only fatpacking for release. This way the code is maintainable and reusable.
If you wish to use ELM assignments in your own project you can now use the classes found in `lib/`.

Please make sure to cite the original ELM authors when using mELM results:

       Holger Dinkel, Kim Van Roey, Sushama Michael, Norman E. Davey,
       Robert J. Weatheritt, Diana Born, Tobias Speck, Daniel Krueger,
       Gleb Grebnev, Marta KubaX, Marta Strumillo, Bora Uyar,
       Aidan Budd, Brigitte Altenberg, Markus Seiler, Lucia B. Chemes,
       Juliana Glavina, Ignacio E. Sanchez, Francesca Diella,
       and Toby J. Gibson (2014)
       The eukaryotic linear motif resource ELM: 10 years and counting
       Nucl. Acids Res. 42(D1): D259-D266
       doi:10.1093/nar/gkt1047

If you have used mELM with ANCHOR predictions please cite the following:

       Balint Meszaros, Istvan Simon and Zsuzsanna Dosztanyi (2009)
       Prediction of Protein Binding Regions in Disordered Proteins
       PLoS Comput Biol 5(5): e1000376.
       doi:10.1371/journal.pcbi.1000376

Usage
=====

        melm [h,v,u,U,c,i,a,G,E,X,m,n,C,P,l,t,M,D,d,g] <SEQ FILES...>
           -h, --help
               this message
           -v, --verbose
               receive information about what melm is doing
           -u, --update
               download and start using the latest class definitions from ELM.eu.org
           -U, --upgrade
               upgrade mELM to the latest version from GitHub.com/MattOates/melm
           -c, --list-classes
               get TSV output for the current ELM classes library melm is using
           -i, --list-instances
               get TSV output of all the ELM instances held in the melm cache
           -a, --assign
               do not mask sequences, instead output a TSV format of all the ELM assignments made per sequence
           -G, -GFF3
               produce GFF3 output when using --assign, useful if you wish to add this assignment to a genome browser or similar
           -E <>, --max-class-expect=<>
               filter out ELM classes based on their annotated expectation, bigger E means allow for more common motifs
           -X, --hard-mask
               when masking use replacement 'x' characters instead of just making lower case (soft masking)
           -m, --mask-motifs
               instead of masking sequence free of motifs mask the motif rich regions
           -n <>, --num-elms=<>
               threshold number of elms per residue assigned to mask on
           -C <>, --min-elm-complexity=<>
               set a minimum limit of complexity rate to assign an ELM, larger number means more complex motifs included
           -P <>, --max-elm-probability=<>
               set the maximum ELM probability, if a short motif is found with high probability its perhaps too common to be a true positive
           -l, --logic-filter
               turn on ELM instance-logic filtering, if there are known false positive ELMs of this class matching the assigned sequence exclude from results
           -t <>, --type=<>
               only include ELM classes of a given type: CLV (cleavage sites), DEG (degradation sites), DOC (docking sites), LIG (ligand binding sites), MOD (ptm sites), TRG (targeting sites)
           -M, --morf-filter
               if ANCHOR is installed use that to filter ELM output based on those having shared coverage with a predicted MoRF region
           -D, --disorder-filter
               if ANCHOR is installed use that to filter ELM output based on ANCHOR's IUPred disorder prediction, only include ELMs that fall within disordered regions
           -d <>, --anchor-datapath=<>
               provide the location of the anchor data path, mELM will otherwise assume it's in the same directory as your anchor binary
           -g <>, --go-filter=<>
               turn on GO filtering, only show results for ELM classes that have been associated with the GO ID specified
           -o <>, --organism-filter=<>
               turn on organism filtering, only show results for ELM instances that have been observed in a given organism

Example Use Cases
=================

Soft mask regions of a protein that do not contain any plausible ELM motifs

           melm human_proteins.fa > human_proteins_motifs.fa

Soft mask all regions of a protein that contain less than two linear motif classes, accounting for known false positives

           melm --num-motifs=2 --logic-filter human_proteins.fa > human_proteins_motif_rich.fa

Mask all PTM sites with 'X' in a set of seqences

           melm --hard-mask --mask-motifs --type=MOD human_proteins.fa > human_no_ptms.fa

Get a GFF3 file for a whole genomes worth of protein annotations

           melm --assign --GFF3 human_proteins.fa > human_motifs.gff3

Get another GFF3 file but this time be strict on assignment to those active in native disordered state from the nucleus
using GO filtering for GO ID 0005634 "nucleus"

           melm --assign --GFF3 --logic-filter --disorder-filter --go-filter=0005634 --organism-filter=sapiens human_proteins.fa > disordered_hiqual_human_nucleus_motifs.gff3

Get the latest ELM classes library for use in another script or by yourself

           melm --update --list-classes

Get the latest ELM instances library for use in another script or by yourself

           melm --update --list-instances

License
=======

           melm - Mask and assign ELM motifs in protein sequence libraries
           (C) 2014-2022  Dr Matt E. Oates

           This program is free software: you can redistribute it and/or modify
           it under the terms of the GNU Affero General Public License as
           published by the Free Software Foundation, either version 3 of the
           License, or (at your option) any later version.

           This program is distributed in the hope that it will be useful,
           but WITHOUT ANY WARRANTY; without even the implied warranty of
           MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
           GNU Affero General Public License for more details.

           You should have received a copy of the GNU Affero General Public License
           along with this program.  If not, see <http://www.gnu.org/licenses/>.



CHANGES.md
###2019-05-17
   * Updated timeout so that downloads from ELM.eu.org are more likely to succeed.
   * Cut a new maintenance release 1.4.3

###2016-07-05
    * Majority of refactor complete

###2016-06-25
   * Huge refactor into split packages allowing more ELM based tools, releases now are fatpack

###2016-05-23
   * Removed bitrot due to changes at ELM.eu.org and enabled support for v1.4 of ELM

###2014-09-10
   * Allowed for the script to upgrade itself from GitHub and made it a bit more friendly to use

###2014-09-09
   * Integrated ANCHOR predictions if this is installed on the system (independent of OS)
   * Limit motif assignment to predicted binding regions (MoRF)
   * Limit motif assignment to predicted disordered regions

###2014-09-08
   * ELM instance downloading
   * False Positive logic filtering based on ELM instances library.
   * GFF3 and assignment output

###2014-09-05
   * Masking with only specific ELM classes included based on expect value and motif type

###2014-08-30
   * First features added.



Build.PL
use strict;
use warnings;
use Module::Build;
 
use 5.008;


my $builder = Module::Build->new(
    module_name         => 'ELM',
    license             => 'lgpl_3',
    dist_author         => 'Matt Oates <mattoates@gmail.com>',
    dist_abstract       => 'CLI tool for assigning Eukaryotic Linear Motifs from the ELM.eu.org database',
    create_makefile_pl  => 0,     #'traditional',
    script_files        => 'bin/melm',
    create_readme       => 0,
    requires            => {
        'LWP::UserAgent' => '6.15',
        'BioPerl'        => '1.7.2',
        'Class::Tiny'    => '1.004',
        'Pod::Usage'     => '1.69'
    },
    build_requires      => {
        'Test::More'                      => '0.47',
        'Test::Pod::Coverage'             => '1.10',
        'Test::Pod'                       => '1.51',
        'Devel::Cover'                    => '1.23',
        'Devel::Cover::Report::Coveralls' => '0.11'
    },
);
 
$builder->create_build_script();



lib/ELM.pm
package ELM v1.4.3;
=encoding UTF-8
=head1 NAME

ELM - Class to do analysis with the ELM regex library

=head1 VERSION

Version v1.4.3

=cut

use v5.20.0;
use strict;
use warnings;
no warnings 'experimental::signatures';
use feature 'signatures';

use ELM::Library;
use ELM::Utils;
use ELM::Anchor;
use ELM::AminoAcids 'score';
use ELM::Calc 'runencode';

use Class::Tiny qw(
    type 
    max_class_expect
    min_elm_complexity
    max_elm_probability
    morf_filter
    disorder_filter
    logic_filter
    go_filter
    organism_filter), {
    num_elms_threshold => sub { 1 },
	library => sub { ELM::Library->new() },
	anchor => sub { ELM::Anchor->new() },
};

=head1 SYNOPSIS

This class is responsible for all high level analysis that can be performed with ELM

    use ELM;

    my $elm = ELM->new();
    ...

=head1 METHODS

=head2 assign_elm

Assign a single ELM class to a given sequence

=cut
sub assign_elm($self, $elm, $seq) {
	my $anchor = $self->morf_filter || $self->disorder_filter;

	#Handle any disorder/morf prediction
	my ($morf_regions, $dis_regions, $iupred);
    if ($anchor) {
        ($morf_regions, $anchor, $iupred) = $self->anchor->assign($seq->seq);
        $dis_regions = runencode($iupred);
    }

    return $self->assign($elm, $self->library->elms->{$elm}{regex}, $seq, $morf_regions, $dis_regions);
}

=head2 assign_all_elms

Assign all ELM classes given filter options to a sequence

=cut
sub assign_all_elms($self, $seq) {

	my @assignment;
    foreach my $elm (keys $self->library->elms->%*) {

    	#Ignore those ELM classes that are below the user defined probability threshold
		if ($self->max_class_expect) {
		    next if $self->library->elms->{$elm}{probability} >= $self->max_class_expect;
		}

		#Ignore all the ELM classes but the type requested
		if ($self->type) {
		    next if $self->library->elms->{$elm}{type} ne $self->type;
		}

        push @assignment, $self->assign_elm($elm, $seq);
    }

    return @assignment;
}

=head2 assign

Take a regular expression and return all of the ([start,end,substr]...) matched results

=cut
sub assign($self, $elm_name, $regex, $string, $morf_regions, $dis_regions) {
    my @ret;
    my ($start,$end,$seq,$prob,$entropy,$entrorate);
    while ($string =~ /($regex)/g) {
        ($start,$end,$seq) = ($-[0]+1, $+[0], $&);
        if ($self->logic_filter) {
            next unless $self->_logic_filter_ok($elm_name,$seq,logic => $self->logic_filter);
        }
        ($prob, $entropy, $entrorate) = score($seq);
        if ($self->max_elm_probability) {
            next if $prob > $self->max_elm_probability;
        }
        if ($self->min_elm_complexity) {
            next if $entrorate < $self->min_elm_complexity;
        }
        if ($self->morf_filter) {
            next unless any_overlap($start,$end,$morf_regions);
        }
        if ($self->disorder_filter) {
            next unless any_overlap($start,$end,$dis_regions);
        }
        if ($self->go_filter and $self->library->go_terms_version) {
            next unless $self->_go_filter_ok($elm_name);
        }
        if ($self->organism_filter) {
            next unless $self->_organism_filter_ok($elm_name);
        }
        push @ret, [$elm_name, $start, $end, $seq, $prob, $entropy, $entrorate];
    }
    if (@ret < 1) {
        return;
    } else {
        return \@ret;
    }
}

=head2 collapse_assignment

Collapse multiple assignments into a single list of disjoint ([start,end]...) and per residue counts of collapsed data

=cut
sub collapse_assignment($self,$sequence,$assignments,%opt) {

    #Create a list of assignment counts per residue
    my @ass = (0) x length $sequence;
    foreach my $motif (map {@$_} @$assignments) {
        $ass[$_-1]++ for $motif->[1]..$motif->[2];
    }

    #Invert the assignment counts, i.e. mark everything that wasn't asigned as being assigned
    #so [5,3,2,1,1,1,2,0,0,3] becomes [0,0,0,0,0,0,0,1,1,0]
    if ($opt{invert}) {
        $_ = abs +($_ >= 1) - 1 for @ass;
    }

    #Runlength encode the assignment flags to get paired assignments again
    #return (\runencode(\@ass), \@ass);
    return runencode(\@ass, threshold => $self->num_elms_threshold);
}

=head2 mask

Mask a string hard/soft based on a list of start,end pairs [[start,end],...]

=cut
sub mask($self,$sequence,$assignments,%opt) {
    foreach my $assignment (@$assignments) {
        if ($opt{hard_mask}) {
            substr($sequence,$assignment->[0]-1,1+$assignment->[1]-$assignment->[0]) = 'x' x (1+$assignment->[1]-$assignment->[0]);  
        } else {
            substr($sequence,$assignment->[0]-1,1+$assignment->[1]-$assignment->[0]) = lc substr($sequence,$assignment->[0]-1,1+$assignment->[1]-$assignment->[0]);  
        }
    }
    $sequence;
}

sub _logic_filter_ok($self, $elm_name, $seq, %opt) {
    my %elms = %{ $self->library->elms };
    $opt{logic} //= 'FP'; #Default to removing only False Positives
    my %filters = map {$_->{seq} => undef} grep {$_->{logic} = $opt{logic}} @{$elms{$elm_name}{instances}};
    return (not exists $filters{$seq})?1:0;
}

sub _go_filter_ok($self, $elm_name) {
     my %elms = %{ $self->library->elms };
     return 0 != scalar grep {$self->go_filter eq $_->{go_id}} @{$elms{$elm_name}{go_terms}};
}

sub _organism_filter_ok($self, $elm_name) {
     # At least one true positive (TP) example from a substr match to the organism text must exit
     my %elms = %{ $self->library->elms };
     return 0 != scalar grep {$_->{logic} eq 'TP' and index(fc $_->{organism}, fc $self->organism_filter ) != -1} @{$elms{$elm_name}{instances}};
}

=head1 AUTHOR

Matt Oates, C<< <mattoates at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<mattoates@gmail.com>, or through
GitHub issues at L<https://github.com/MattOates/melm/issues>.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ELM


You can also look for information at:

=over 4

=item * GitHub L<https://github.com/MattOates/melm>

=back


=head1 ACKNOWLEDGEMENTS

Please make sure to cite the original ELM authors when using mELM results:
    Holger Dinkel, Kim Van Roey, Sushama Michael, Norman E. Davey, 
    Robert J. Weatheritt, Diana Born, Tobias Speck, Daniel Krüger, 
    Gleb Grebnev, Marta Kubań, Marta Strumillo, Bora Uyar, 
    Aidan Budd, Brigitte Altenberg, Markus Seiler, Lucía B. Chemes,
    Juliana Glavina, Ignacio E. Sánchez, Francesca Diella, 
    and Toby J. Gibson (2014)
    The eukaryotic linear motif resource ELM: 10 years and counting
    Nucl. Acids Res. 42(D1): D259-D266 
    doi:10.1093/nar/gkt1047

If you have used mELM with ANCHOR predictions please cite the following:
    Bálint Mészáros, István Simon and Zsuzsanna Dosztányi (2009) 
    Prediction of Protein Binding Regions in Disordered Proteins
    PLoS Comput Biol 5(5): e1000376. 
    doi:10.1371/journal.pcbi.1000376

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Matt Oates.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut

1;


lib/ELM/Calc.pm
package ELM::Calc v1.4.3;
require Exporter;
=encoding UTF-8
=head1 NAME

ELM::Calc - Functions for calculating sequence assignment specific tasks

=head1 VERSION

Version v1.4.3

=cut

use v5.20.0;
use strict;
use warnings;
no warnings 'experimental::signatures';
use feature 'signatures';

our @ISA = qw(Exporter);
our @EXPORT_OK = qw(coverage any_overlap runencode);

=head1 SYNOPSIS

This module exports functions for calculating common sequence assignment tasks

=head1 FUNCTIONS

=head2 coverage

Efficiently calculate the coverage between two number line segments

=cut
sub coverage($a, $b, $x, $y) {
    return max(min((1+$b-$a),(1+$y-$x),(1+abs($b-$a)+abs($y-$x))-(max($a,$b,$x,$y)-min($a,$b,$x,$y))),0);
}

=head2 any_overlap

Report if there is any overlap between a single defined region and a set of regions

=cut
sub any_overlap($start, $end, $pairs) {
    foreach my $pair (@$pairs) {
        return 1 if coverage($start,$end,$pair->[0],$pair->[1]);
    }
    return 0;
}

=head2 runencode

Take an arrayref of numbers or a delimited string and returns a list of start-end pairs for runs of values above a threshold

=cut
sub runencode($probs, %opts) {

        unless (defined $probs) {
            warn "undef passed to runencode!";
            return;
        }

        #Some default optional parameters
        $opts{delimiter} //= ','; 
        $opts{threshold} //= 0.5;
        $opts{cmp} //= sub {$_[0] >= $_[1]};

        #If a string was passed split it into values
        $probs = [split /$opts{delimiter}/, $probs] if (ref $probs ne 'ARRAY');
        my @ranges = ();
        my $index = 1;
        my $start = undef;
        my $end = undef;

        foreach my $prob (@$probs) {
                if ($opts{cmp}->($prob, $opts{threshold})) {
                        $start = $index unless defined $start;
                } else {
                        $end = $index-1;
                        push @ranges, [$start, $end] if defined $start;
                        $start = undef;
                }
                $index++;
        }

        #Deal with the edge case where there is a region at the end of the sequence
        if ($probs->[-1] >= 0.5) {
                        $end = $index-1;
                        push @ranges, [$start, $end] if defined $start;
        }

        return \@ranges;
}

=head1 AUTHOR

Matt Oates, C<< <mattoates at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<mattoates@gmail.com>, or through
GitHub issues at L<https://github.com/MattOates/melm/issues>.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ELM::Calc


You can also look for information at:

=over 4

=item * GitHub L<https://github.com/MattOates/melm>

=back


=head1 ACKNOWLEDGEMENTS

Please make sure to cite the original ELM authors when using mELM results:
    Holger Dinkel, Kim Van Roey, Sushama Michael, Norman E. Davey, 
    Robert J. Weatheritt, Diana Born, Tobias Speck, Daniel Krüger, 
    Gleb Grebnev, Marta Kubań, Marta Strumillo, Bora Uyar, 
    Aidan Budd, Brigitte Altenberg, Markus Seiler, Lucía B. Chemes,
    Juliana Glavina, Ignacio E. Sánchez, Francesca Diella, 
    and Toby J. Gibson (2014)
    The eukaryotic linear motif resource ELM: 10 years and counting
    Nucl. Acids Res. 42(D1): D259-D266 
    doi:10.1093/nar/gkt1047

If you have used mELM with ANCHOR predictions please cite the following:
    Bálint Mészáros, István Simon and Zsuzsanna Dosztányi (2009) 
    Prediction of Protein Binding Regions in Disordered Proteins
    PLoS Comput Biol 5(5): e1000376. 
    doi:10.1371/journal.pcbi.1000376

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Matt Oates.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut

1;

lib/ELM/Utils.pm
package ELM::Utils v1.4.3;
require Exporter;
=encoding UTF-8
=head1 NAME

ELM::Utils - Common helper functions

=head1 VERSION

Version v1.4.3

=cut

use v5.20.0;
use strict;
use warnings;
no warnings 'experimental::signatures';
use feature 'signatures';

our @ISA = qw(Exporter);
our @EXPORT_OK = qw(get_www);

use LWP::UserAgent;

=head1 SYNOPSIS

This module exports common helper functions

=head1 FUNCTIONS

=head2 get_www

Get the source of a webpage

=cut
sub get_www($url) {
    if (not eval { require LWP::UserAgent }) {
        say STDERR "On first execution and when updating melm requires your Perl installation to have LWP::UserAgent to contact ELM servers for data.";
        say STDERR "You might be able to install this module using `cpanm -i LWP::UserAgent`.";
        say STDERR "Another alternative is to download the latest ELM data from the melm repository at: https://raw.githubusercontent.com/MattOates/melm/master/.elm.dat";
        say STDERR "Place the .elm.dat file into your home directory: /Users/username (OSX), /home/username (Linux) or C:\\Users\\username (Windows)";
        exit 1;
    }
    if ($url =~ /^https/ and not eval { require LWP::Protocol::https }) {
        say STDERR 'Sorry $url requires HTTPS support and your Perl does not have LWP::Protocol::https installed.';
        say STDERR 'Try `cpanm -i LWP::Protocol::https` to install this module.';
        say STDERR 'If you are running this code from an official release distribution please raise a GitHub issue.';
        exit 1;
    }
    #Decided to localise this and just handle checking for deps properly
    my $ua = LWP::UserAgent->new;
    $ua->timeout(600); # Wait at least 10 minutes, ELM can be very slow to respond

    my $response = $ua->get($url);
    if ($response->is_success) {
       return $response->decoded_content;
    } else {
        die "Failed whilst contacting $url: " . $response->status_line;
    }
}

=head1 AUTHOR

Matt Oates, C<< <mattoates at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<mattoates@gmail.com>, or through
GitHub issues at L<https://github.com/MattOates/melm/issues>.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ELM::Utils


You can also look for information at:

=over 4

=item * GitHub L<https://github.com/MattOates/melm>

=back


=head1 ACKNOWLEDGEMENTS

Please make sure to cite the original ELM authors when using mELM results:
    Holger Dinkel, Kim Van Roey, Sushama Michael, Norman E. Davey, 
    Robert J. Weatheritt, Diana Born, Tobias Speck, Daniel Krüger, 
    Gleb Grebnev, Marta Kubań, Marta Strumillo, Bora Uyar, 
    Aidan Budd, Brigitte Altenberg, Markus Seiler, Lucía B. Chemes,
    Juliana Glavina, Ignacio E. Sánchez, Francesca Diella, 
    and Toby J. Gibson (2014)
    The eukaryotic linear motif resource ELM: 10 years and counting
    Nucl. Acids Res. 42(D1): D259-D266 
    doi:10.1093/nar/gkt1047

If you have used mELM with ANCHOR predictions please cite the following:
    Bálint Mészáros, István Simon and Zsuzsanna Dosztányi (2009) 
    Prediction of Protein Binding Regions in Disordered Proteins
    PLoS Comput Biol 5(5): e1000376. 
    doi:10.1371/journal.pcbi.1000376

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Matt Oates.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut

1;

lib/ELM/Library.pm
package ELM::Library v1.4.3;
=encoding UTF-8
=head1 NAME

ELM::Library - Class to hold the ELM regex library

=head1 VERSION

Version v1.4.3

=cut

use v5.20.0;
use strict;
use warnings;
use autodie;
no warnings 'experimental::signatures';
use feature 'signatures';

use ELM::Utils 'get_www';

use Cwd qw/abs_path/;
use Data::Dumper;
#Make eval of dump simpler
$Data::Dumper::Terse=1;

use Class::Tiny qw(
    elms
    go_terms_version
    classes_version
    instances_version),{
    elm_lib_path => sub { $ENV{HOME} . '/.elm.dat' }
};

=head1 SYNOPSIS

This class is responsible for all high level operations on creating and maintaining a local ELM library.
Data is automatically downloaded from elm.eu.org and can be updated programatically.

To create an ELM::Library explicitly.

    use ELM::Library;

    my $lib = ELM::Library->new();
    ...

=head1 METHODS

=head2 BUILD

Load the ELM classes when we instantiate

=cut
sub BUILD($self, $args) {
    $self->load_elm_classes;
}

=head2 exists

Check to see if the library exists

=cut
sub exists($self) {
    return -e $self->elm_lib_path;
}

=head2 load_elm_classes

Load the cached ELM classes/instances data

=cut
sub load_elm_classes($self) {
    if ($self->exists) {
        #Load the cached ELM classes/instances data
        local $/; #slurp
        open my $elm_fh, '<', $self->elm_lib_path;
        my ($classes_version, $instances_version, $classes, $go_terms_version) =  @{ eval <$elm_fh> };
        close $elm_fh;
        $self->elms($classes);
        $self->classes_version($classes_version);
        $self->instances_version($instances_version);
        if ($go_terms_version) {
            $self->go_terms_version($go_terms_version);
        } else {
            say STDERR "Local cache did not contain GO terms, you might want to update to v1.4.3"
        }
    } else {
        say STDERR "Could not find an ELM library at " . $self->elm_lib_path . " so auto-fetching a fresh one for the first time";
        $self->update;
    }
}

=head2 update

Update the ELM definitions from elm.eu.org

=cut
sub update($self) {
    say STDERR "Updating ELM classes file " . $self->elm_lib_path;
    $self->_update_elm_classes();
    $self->_update_elm_instances();
    $self->_update_go_terms();
    open my $elm_fh,'>', $self->elm_lib_path;
    say $elm_fh Dumper([$self->classes_version, $self->instances_version, $self->elms, $self->go_terms_version]);
    close $elm_fh;
}

=head2 list_classes

Print out all of the ELM classes in a simple TSV format

=cut
sub list_classes($self) {
    my %elms = %{ $self->elms };
    say "#Cached melm data for version $self->classes_version of the ELM classes library";
    say "#" . join "\t", 'Accession', 'Type', 'Name', 'Description', 'Regex', 'Expectation';
    foreach my $elm (keys %elms) {
        say "$elms{$elm}{accession}\t$elms{$elm}{type}\t$elm\t$elms{$elm}{description}\t$elms{$elm}{regex}\t$elms{$elm}{probability}";
    }
}

=head2 list_instances

Print out all of the ELM instances in a simple TSV format

=cut
sub list_instances($self) {
    my %elms = %{ $self->elms };
    say "#Cached melm data for version $self->instances_version of the ELM instances library";
    say "#" . join "\t", 'Accession', 'Name', 'Primary UniProt Accession', 'Start', 'End', 'Sequence', 'Assignment Logic';
    foreach my $elm (keys %elms) {
        foreach my $instance (@{$elms{$elm}{instances}}) {
            say "$instance->{accession}\t$elm\t$instance->{id}\t$instance->{start}\t$instance->{end}\t$instance->{seq}\t$instance->{logic}";
        }
    }
}

#Refresh the cached ELM classes file and populate %elms with latest data
sub _update_elm_classes($self) {
    my $classes_version;
    $self->elms({}); #Wipe out existing elms
    my $class_tsv = get_www('http://elm.eu.org/elms/elms_index.tsv');
    #Each record looks like:
    #Pre 1.4
    #Accession  ELMIdentifier   Description Regex   Probability #Instances  #Instances_in_PDB
    #Post 1.4
    #Accession ELMIdentifier FunctionalSiteName Description Regex Probability #Instances #Instances_in_PDB
    #ELME000080"    "TRG_PEX_1" "Wxxx[FY] motifs present in N-terminal half of Pex5 bind to Pex13 and Pex14 at peroxisomal and glycosomal membranes to facilitate entrance of PTS1 cargo proteins into the organellar lumen. =>"W...[FY] =>"0.000222625 =>"27"  "1"
    foreach my $record (split /\n/, $class_tsv) {
        (undef,$classes_version) = split /: /, $record if ($record =~ /^#ELM_Classes_Download_Version/);
        next if $record =~ /^(#|"Accession)/; #Ignore the header
        $record =~ s/"//g;
        my ($elm_id,$elm_name,$name,$description,$regex,$probability,$instances,$instances_in_pdb) = split /\t/, $record;
        my ($type) = split /_/, $elm_name;
        $self->elms->{$elm_name} = {accession => $elm_id, elm_name => $elm_name, site => $name, type => $type, description => $description, regex => $regex, probability => $probability};
    }
    $self->classes_version($classes_version);
    say STDERR "Updated ELM library to version $classes_version";
}

#Get the sequence for an instance from ELM
sub _get_instance_seqs($self) {
    say STDERR "Downloading instance protein sequences from ELM.";
    my $fasta = get_www("http://elm.eu.org/instances.fasta?q=*");
    my %sequences;
    #Hard assumption, ELM gives out sequences as a single line (appears true)
    foreach my $record (split />/, $fasta) {
        next if $record =~ /^\s*$/;
        my ($id,$seq) = split /\n/, $record;
        my (undef, $up_id) = split /\|/, $id;
        $sequences{$up_id} = $seq;
    }
    return %sequences;
}

#Refresh the cached ELM instances data
sub _update_elm_instances($self) {
    my $instances_version;
    my %instance_logic = ('false positive' => 'FP','true negative' => 'TN','true positive', => 'TP', 'unknown' => 'U');
    my $instances_tsv = get_www('http://elm.eu.org/instances.tsv?q=*');
    my %uniprot_sequences = $self->_get_instance_seqs();
    #Each record looks like:
    #Accession, ELMType, ELMIdentifier, ProteinName, Primary_Acc, Accessions, Start, End, References, Methods, InstanceLogic, PDB, Organism
    #"ELMI000471"   "LIG"   "LIG_SH2_SRC"   "SRC_HUMAN" "P12931"    "P12931 E1P5V4 Q76P87 Q86VB9 Q9H5A8"    "530"   "533"   "10360179"  "x-ray crystallography" "true positive" ""  "Homo sapiens"
    foreach my $record (split /\n/, $instances_tsv) {
        (undef, $instances_version) = split /: /, $record if ($record =~ /^#ELM_Instance_Download_Version/);
        next if $record =~ /^(#|"Accession)/; #Ignore the header
        $record =~ s/"//g;
        $record =~ s/\r//g;
        my ($elm_id, $type, $elm_name, $protein_name, $up_id, $alt_up_id, $start, $end, $references, $methods, $logic, $pdb, $organism) = split /\t/, $record;
        my $seq = substr( $uniprot_sequences{$up_id}, $start-1, 1+$end-$start );
        push @{$self->elms->{$elm_name}{instances}}, {accession => $elm_id, id => $up_id, start => $start, end => $end, logic => $instance_logic{$logic}, seq => $seq, organism => $organism};
    }
    say STDERR "Instances data updated to version $instances_version.";
    $self->instances_version($instances_version);
}

#Refresh the cached GO term data
sub _update_go_terms($self) {
    my $go_terms_version;
    my $go_term_tsv = get_www('http://elm.eu.org/goterms.tsv');

    foreach my $record (split /\n/, $go_term_tsv) {
        (undef, $go_terms_version) = split /: /, $record if ($record =~ /^#ELM_GOTERMs_Download_Version/);
        next if $record =~ /^(#|"ELM)/; #Ignore the header
        $record =~ s/"//g;
        $record =~ s/\r//g;
        my ($elm_name, $go_id, $go_term) = split /\t/, $record;
        $go_id =~ s/^GO://;
        push @{$self->elms->{$elm_name}{go_terms}}, {go_id => $go_id, go_term => $go_term};
    }
    say STDERR "GO data updated to version $go_terms_version.";
    $self->go_terms_version($go_terms_version);
}

=head1 AUTHOR

Matt Oates, C<< <mattoates at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<mattoates@gmail.com>, or through
GitHub issues at L<https://github.com/MattOates/melm/issues>.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ELM::Library


You can also look for information at:

=over 4

=item * GitHub L<https://github.com/MattOates/melm>

=back


=head1 ACKNOWLEDGEMENTS

Please make sure to cite the original ELM authors when using mELM results:
    Holger Dinkel, Kim Van Roey, Sushama Michael, Norman E. Davey, 
    Robert J. Weatheritt, Diana Born, Tobias Speck, Daniel Krüger, 
    Gleb Grebnev, Marta Kubań, Marta Strumillo, Bora Uyar, 
    Aidan Budd, Brigitte Altenberg, Markus Seiler, Lucía B. Chemes,
    Juliana Glavina, Ignacio E. Sánchez, Francesca Diella, 
    and Toby J. Gibson (2014)
    The eukaryotic linear motif resource ELM: 10 years and counting
    Nucl. Acids Res. 42(D1): D259-D266 
    doi:10.1093/nar/gkt1047

If you have used mELM with ANCHOR predictions please cite the following:
    Bálint Mészáros, István Simon and Zsuzsanna Dosztányi (2009) 
    Prediction of Protein Binding Regions in Disordered Proteins
    PLoS Comput Biol 5(5): e1000376. 
    doi:10.1371/journal.pcbi.1000376

=head1 LICENSE AND COPYRIGHT

Copyright 2022 Matt Oates.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


=cut

1;